packages/hardware-trezor/src/TrezorKeyAgent.ts

```diff
@@ -0,0 +1,180 @@
+        popup: communicationType !== CommunicationType.Node && !silentMode
+      });
+      return true;
+    } catch (error: any) {
+      if (error.code === 'Init_AlreadyInitialized') return true;
+      throw transportTypedError(error);
+    }
+  }
+
+  /**
+   * @throws AuthenticationError
+   */
+  static async checkDeviceConnection(): Promise<Features> {
+    const deviceFeatures = await TrezorConnect.getFeatures();
+    if (!deviceFeatures.success) {
+      throw transportTypedError(new errors.TransportError('Failed to get device', deviceFeatures.payload));
+    }
+    if (deviceFeatures.payload.model !== 'T') {
+      throw transportTypedError(
+        new errors.TransportError(`Trezor device model "${deviceFeatures.payload.model}" is not supported.`)
+      );
+    }
+    return deviceFeatures.payload;
+  }
+
+  /**
+   * @throws AuthenticationError
+   */
+  static async getXpub({ accountIndex }: GetTrezorXpubProps): Promise<Crypto.Bip32PublicKeyHex> {
+    await TrezorKeyAgent.checkDeviceConnection();
+    const derivationPath = `m/${CardanoKeyConst.PURPOSE}'/${CardanoKeyConst.COIN_TYPE}'/${accountIndex}'`;
+    const extendedPublicKey = await TrezorConnect.cardanoGetPublicKey({
+      path: derivationPath,
+      showOnTrezor: true
+    });
+    if (!extendedPublicKey.success) {
+      throw transportTypedError(
+        new errors.TransportError('Failed to export extended account public key', extendedPublicKey.payload)
+      );
+    }
+    return Crypto.Bip32PublicKeyHex(extendedPublicKey.payload.publicKey);
+  }
+
+  /**
+   * @throws AuthenticationError
+   */
+  static async createWithDevice(
+    { chainId, accountIndex = 0, trezorConfig }: CreateTrezorKeyAgentProps,
+    dependencies: KeyAgentDependencies
+  ) {
+    const isTrezorInitialized = await TrezorKeyAgent.initializeTrezorTransport(trezorConfig);
+    const extendedAccountPublicKey = await TrezorKeyAgent.getXpub({ accountIndex });
+    return new TrezorKeyAgent(
+      {
+        accountIndex,
+        chainId,
+        extendedAccountPublicKey,
+        isTrezorInitialized,
+        knownAddresses: [],
+        trezorConfig
+      },
+      dependencies
+    );
+  }
+
+  async signTransaction({ body }: Cardano.TxBodyWithHash): Promise<Cardano.Signatures> {
+    const scope = new ManagedFreeableScope();
+    try {
+      await this.isTrezorInitialized;
+      const cslTxBody = coreToCml.txBody(scope, body);
+      const trezorTxData = await txToTrezor({
+        accountIndex: this.accountIndex,
+        cardanoTxBody: body,
+        chainId: this.chainId,
+        cslTxBody,
+        inputResolver: this.inputResolver,
+        knownAddresses: this.knownAddresses
+      });
+
+      const result = await TrezorConnect.cardanoSignTransaction(trezorTxData);
+      if (!result.success) {
+        throw new errors.TransportError('Failed to export extended account public key', result.payload);
+      }
+
+      const signedData = result.payload;
```

Comment:
This will help catch any mapping errors without having to submit the signed transactions in tests (`hash` available in the arg object Cardano.TxBodyWithHash)
```suggestion
      const signedData = result.payload;
      if (signedData.hash !== hash) {
        throw new errors.HwMappingError('Trezor computed a different transaction id');
      }
```

Reply:
Resolved in d045670ed2eb5eee8efa1acaf97823ebb7676922